### DISCUSSION TITLE
Increasing expressiveness of the CIP numbering system

### DISCUSSION THREAD

**Discussion by felipeargento** [cartesi/cips#11] [Ideas]
Hello,

I was thinking that numbering CIPs sequentially might not be the best way to do it. The worry is that as the number of CIPs grow it will be harder and harder to keep track of them all.

At the same time, groups that are specialized in an area might only be interested (or specially interested) in CIPs related that area.

At the moment, we have four types of CIPs:
- Core
- Application
- Meta
- Informational

In the [explanation](https://github.com/cartesi/cips/blob/cip3/cips/cip-1.md#cip-types) of each type, @fcecin list some examples of subtypes. For instance, an Application CIP might be about methodologies, best practices, interfaces. A Core CIP might be about interoperability, emulator, onchain code etc.

I started drafting a CIP proposing a new way of enumerating CIPs that can express which area of Cartesi's ecosystem/infrastructure they plan to improve.

My initial thoughts is that the number of a CIP proposal should be `1 digit for type + 2 digits for subtype + 3 digits for proposal number`. The "proposal number" would be the number of proposals created on that `type + subtype + 1`.

If we decide, for instance, that Core is type 1 and that onchain code is subtype 2, we'd have that the first CIP proposal for an onchain improvement would be: `102001`.

Another proposal, to deal with the asynchronicity, is for the number to be: 
`1 digit for type + 2 digits for subtype + PR number`.
In that case, the CIP aforementioned would be number `102[PR number]`

If we go with something similar to the first proposal, we need to decide when the CIP gets its number. Otherwise we might end up with some race conditions :) 

I might be using too many digits, but would love to hear your thoughts on this.

**Comment by fcecin** [comment]
I like it!

I'm not sure whether PR is really better than a manual sequence, even if you have to disambiguate manually sometimes. I would guess using Github PR numbers is the better option (given that Github is not going anywhere).

And even if we don't get to 10 subtypes, I like the extra zero on the subtype as a separator.

We could also use the first letter of the main type as the CIP "number" (i.e. code). There's no fundamental reason, I think, that the CIP code has to be an actual number.

**Reply by felipeargento** [reply]
Indeed disambiguating manually shouldn't be too cumbersome, specially if the branches are properly named. The only issue I see with github is that we only get a number once the PR is open - but before I guess we can use the title to refer to it :)

About the subtypes as a separator - indeed it looks better. I thought about using an explicit separator like `-`, but thought that would make it awkward to say the number. 

Using the first letter of the main type is an interesting proposal, it would look good and be even easier to identify. The only concern is the possibility of having duplicates, but assuming that we're already accepting a max of 9 types I am sure we can find enough synonyms for new types to manipulate it's first letter.

I think I am now leaning towards github number and first letter of the type.

**Reply by guidanoli** [reply]
I like the idea of using letters in CIP identifiers. It helps to distinguish the type of the CIP (for example, a Core Emulator CIP could have a "CE" identifier) and the number of the CIP (which could still be the PR number that GitHub automatically assigns). One benefit of doing so is that we don't need to pad the CIP number with zeros, so we don't assume an upper bound for the number of CIPs per type. Some examples:

* CIP-CE33 (Core Emulator CIP whose PR number is 33)
* CIP-AI42 (Application Interface CIP whose PR number is 42)

**Reply by fcecin** [reply]
If we go with two letters or more as part of the "number", we get a subtle linguistic side benefit that allows us to drop the "CIP-" prefix from naming the things.

I.e., you can refer directly to the "AI42" or the "CE33", using them as nouns in writing or speech, because now you have a syllable of sorts, so you can drop the "CIP-" prefix. All of the prefix combinations of type-subtype (which is really a composite type) are now pseudo-syllables (abbreviations, acronyms).

This is also a kind of subtle power play because now "Cartesi" is implied in speech, like when ".com" sites are assumed to be USian domains without the ".us". And, it also makes us say "CIP" less -- an acronym that is being used both for the whole system and for each improvement -- which is good.

**Comment by guidanoli** [comment]
This reminded me of how highways are numbered here in Brazil. [1]
They also contain a prefixing digit that tells something more about them. For example:

* `BR-0xx` are radial highways that come from the capital, where `xx` increases clockwise.
* `BR-1xx` are vertical highways, where `xx` increases from left to right.
* `BR-2xx` are horizontal highways, where `xx` increases from top to bottom.

[1] Departamento Nacional de Infraestrutura de Transportes. Nomenclatura das rodovias federais. Retrieved April 6, 2022, from https://www.gov.br/dnit/pt-br/rodovias/rodovias-federais/nomeclatura-das-rodovias-federais/nomeclatura-das-rodovias-federais-1

**Comment by fcecin** [comment]
As the champion of this proposal, @felipeargento, once the new name and numbering scheme is decided here, you can submit a PR to directly patch CIP-1.

**Reply by felipeargento** [reply]
I was wondering about that, would we do backwards renaming?

**Reply by fcecin** [reply]
Good point, yes, we would do that, especially as we are in the beginning so that's going to be manageable.

**Comment by felipeargento** [comment]
Alright, I fully support the idea of using letters as the identifiers...looks much better! I'd love to have @augustoteixeira take on this, since he is (for now) our most prolific CIP writer.

What I am questioning now is if we should go on an effort to define/categorize the subtypes or if we let them rise naturally. What do you think?

**Reply by guidanoli** [reply]
For now, if some subtypes are already clear to us, then I think we can define identifiers for those.
And, in the future, if the need for new subtypes emerges, then we could make room for those too.

**Reply by fcecin** [reply]
We can have a dedicated section in CIP-1 with the full list of types and subtypes, and update it as needed.

**Reply by guidanoli** [reply]
@fcecin I too think we should concentrate this information somewhere like CIP-1.

**Comment by augustoteixeira** [comment]
I really like the idea and how it evolved. I specially like the `CIP-CE-33` proposal, with:
- `C` standing for the type, in this case Core
- `E` standing for sub-type, in this case Emulator
- The `CIP` being dropped from speech and text when the context is clear.

The only question I still have concerns the numbering:
- if we use a manual numbering, people should coordinate (I am preparing `CE-12` at the same time as Felipe is also preparing `CE-12`, because they are the next number available).
- if we use the PR number, this forces us to create a PR and it could be difficult to refer to CIP`s during the discussion phase (which can take long), because the PR is only created after the discussion is over, right?
- if we use the discussion number, it could work, but it has the problem of big gaps, like there will be `CE-3`, `CE-18`... but nothing in between. This problem also occurs with PR numbers.

I have a tendency to prefer the manual numbering. But what are your thoughts?

**Reply by fcecin** [reply]
EIP solves the "no PR" problem by having the CIP file be originally submitted in a PR branch with a short name instead of a number. e.g. "cip-setup-input.md" (this is also what you put on the header, inside the file). It is actually the CIP Editor that's taking care of each CIP that assigns that number, not the author. The latter happens when there's definitely a PR already created to refer to, which is exactly that PR that is created with the cip-file-name.md temporary convention.

Since I don't like unnecessary bureaucracy, I'd rather have the authors number them as best as they can, and if they themselves see a conflict they can resolve it in their own branches. When the Draft CIP is actually committed to the repo from its PR branch, it is done by looking at all the CIP numbers that already exist in the trunk; that's where the CIP actually gets the official number. While it is in a PR branch, the number is fake, tentative, and can even be some short name instead. So I'd let the authors sort the number in their PR branches, and when the CIP Editor is approving the merge to the trunk, it will either accept the number or block it with "oops it looks someone was faster than you".

Finally, the link from the CIP to the discussion forum is more important than the link from the discussion forum to the CIP. The CIP links will change as they start their life as e.g. a google docs, then some cips repo branch, then the cips repo main/trunk branch with the official name/number/code.

EDIT: So I guess what I'm saying is that if your number is crap, the Editor has the power to assign the PR number to it.

EDIT: Or, to be more formal: if the temporary name you choose for your CIP draft in your PR branch happens by accident to be a valid CIP code, it is kept, otherwise it is overriden by whatever the Editor wants, which can be the PR number if that happens to be available.

**Comment by fcecin** [comment]
Also, if anybody else wants to be an official CIP Editor, that would be great. I'd rather be the Meta-Editor, i.e., work on sorting out how the system itself works. And it would be nice to have like 3 editors at least.

**Comment by augustoteixeira** [comment]
I like very much the idea of giving it a short name before the number gets assigned.

Like `CIP-CR-aggregation-encoding` exists until the first PR, at which point it becomes `CIP-CR-32`, because 32 was the number of the PR. Of course it can be referred to as `CR-32` if wanted.

I am ready to finish the discussion with this standard. What about you?

On another notice, I also like to write to the CIP-1 a list of all the sub-categories that already exist. Probably it is only `CR` for now :) So I would probably write a few more just because we know they are needed, right? Like `CE`, `CI` for infrastructure...

**Reply by fcecin** [reply]
I would not add a dash between the letter prefix and the number suffix (i.e. CIP-CR32 or CR32, instead of CR-32, as it can be seen as a whole contiguous code that's a mix of letters and numbers), but that's just my opinion, we can go with either.

The subtype "X" can be reserved to mean "no subtype" or "generic/catch all", so that we get a CIP-CX-200 intead of a CIP-C-200 (variable witdh in the letter prefix might be confusing). CIPs that get the "X" subtype can be renamed later (using the superseded-by mechanic, perhaps) if a suitable subtype later appears. But hopefully a CIP without a proper subtype prompts the creation of the subtype before it is committed to main with the X.

**Reply by fcecin** [reply]
Addendum: some types may be without a subtype. I'm thinking about Informative (I) CIPs. Instead of naming all of them IX, we can name them "IN" instead, if we know the type won't have subtypes.

**Comment by fcecin** [comment]
By questioning the numbering methodology, I think @felipeargento has been promoted to CIP Editor.

Felipe, add yourself to the CIP Editors list in CIP-1!

**Comment by felipeargento** [comment]
I created the [CIP](https://github.com/cartesi/cips/pull/23)! Will also include myself as an Editor in that PR, @fcecin :)
